src/dblib/dblib.c

```diff
@@ -803,6 +803,10 @@ dbsetlname(LOGINREC * login, const char *value, int which)
 	case DBSETSERVERPRINCIPAL:
 		copy_ret = !!tds_dstr_copy(&login->tds_login->server_spn, value_nonull);
 		break;
+	case DBSETENCRYPTION:
+		tds_parse_conf_section(TDS_STR_ENCRYPTION, value_nonull, login->tds_login);
+		return SUCCEED;
```

Comment:
There's no way to return error from `tds_parse_conf_section` but maybe instead of
```c
return SUCCEED;
```
you could do something like
```c
copy_ret = !!login->tds_login->valid_configuration;
```
to detect error in login configuration.

Reply:
I'm not sure about this...
I've checked that regardless of the validity of the `encryption` parameter value `off/request/require` or any other invalid values `login->tds_login->valid_configuration` is `0` before calling `tds_parse_conf_section` and this value does not change after executing `tds_parse_conf_section`.
Let's take a look inside `tds_parse_conf_section` :
```
	} else if (!strcmp(option, TDS_STR_ENCRYPTION)) {
		tds_config_encryption(value, login);
	}
```
I've checked that `login->valid_configuration` is `0` before calling of `tds_config_encryption` and of course after execution of `tds_config_encryption` is still `0` because `tds_config_encryption` does not set `valid_configuration` to `1` anywhere.
Should we set `valid_configuration` to `1` before calling `tds_parse_conf_section` or set it with `tds_config_encryption`, or maybe do you have any other idea to deal with this problem?

Reply:
`valid_configuration` is set initially to 1 (true) and reset (set to false) if any error happens.

Yes, you are right, in this case if we would like to detect the error excluding previous ones we would need to set to 1 and check if it's set to 0 by `tds_parse_conf_section`.

Reply:
I thought that `valid_configuration` should be initially set to `1` but in fact it is not initialized or I can't find it.
I found comment in `dblogin` `/* set default values for loginrec */` but nothing here sets default values for `loginrec` at all. So I thought of calling `tds_init_login` here. Now, we have the `LOGINREC` structure initialized and of course also the `valid_configuration`.
However, all this does not change the fact that assigning the wrong value to `encryption` causes the connection to be established and the default value to be set `require`, assuming `require` is safer than `no`.

Reply:
I think I'm creating more confusion than needed. Reading the code and trying to remember now I understood. The `valid_configuration` is mainly used to detect wrong configuration from the file (like `freetds.conf`), not from user configuration. Time ago there were 2 structures, `TDS_LOGIN` and `TDS_CONNECTION`, first used to store the configuration from user, second to store configuration from file and final configuration (used for connection).
So a library usually:
- creates a `TDS_LOGIN`
- fills the values (like using `dbsetlname`)
- creates a `TDS_CONNECTION`
- fill the `TDS_CONNECTION` from file information (you need server name from `TDS_LOGIN`
- merge configuration from `TDS_LOGIN` to `TDS_CONNECTION`
- connect to server using `TDS_CONNECTION`

The `valid_configuration` was introduces for `TDS_CONNECTION` (now the 2 types points to the same structure), so not to detect errors from user settings (that's the reason code works with `valid_configuration` set to false on `TDS_LOGIN`).

I suggested (probably a wrong suggestion) to use `valid_configuration` (thinking that this flag was used by `TDS_LOGIN` also) to detect a single error for the specific setting but the code indeed is getting more complicated. Looking again at the code I think a more direct error not changing `valid_configuration` would be better.

Reply:
This should simplify the job https://github.com/FreeTDS/freetds/commit/e20a6515f58bdf4a27f937e069fda8bc6ca4116e (already in master)

Reply:
Great! This changes a lot and makes it easier.
Thank you for all the tips and information. Always good to know.